caput/mpiarray.py

```diff
@@ -1333,32 +1333,83 @@ def reshape(self, *shape):
             Reshaped MPIArray as a view of the original data.
         """
 
+        def _check_shapes(
+            current_shape: Tuple[int], new_shape: Tuple[int]
+        ) -> Tuple[int]:
+            """Check that we can reshape one array into another.
+
+            Returns the fully fleshed out shape, or raises a ValueError.
+            """
+
+            actual_size = 1
+            for s in current_shape:
+                actual_size *= s
+
+            wildcard_pos = -1
+
+            new_size = 1
+            for ii, s in enumerate(new_shape):
+                if s == -1 and wildcard_pos < 0:
+                    wildcard_pos = ii
+                elif s == -1 and wildcard_pos >= 0:
+                    raise ValueError(
+                        f"Found two wildcards (-1) in new array shape {new_shape}"
+                    )
+                else:
+                    new_size *= s
+
+            if wildcard_pos >= 0 and actual_size % new_size == 0:
+                # If there was a wildcard then it then we need to check that it would be
+                # a sane value and calculate what it is
+                new_shape = new_shape[:]
+                new_shape[wildcard_pos] = actual_size // new_size
```

Comment:
This won't work if `new_shape` is actually a tuple, it's being given a list when this is called (line 1382). Should either change the type annotation or explicitly cast it to a list in this function.